# **Interacting with OpenCV**

These set of notebooks correspond to the tutorial provided by [ProgrammingKnowldge](https://www.youtube.com/channel/UCs6nmQViDpUw0nuIx9c_WvA) YouTube channel which consists of a series of videos as a way of introducing OpenCV with Python. You can also check the full tutorial (around 9 hours and 23 minutes of length) at the following [link](https://www.youtube.com/watch?v=N81PCpADwKQ).

In this particular notebook, we'll see how to interact with OpenCV, including:

- Drawing shapes on images
- Setting camera parameters and showing datetime information in videos
- Handling mouse events

In [1]:
import cv2
import os

path = 'C:\\Users\\Aristi\\Documents\\Cursos\\OpenCV\\SampleImages'
os.chdir(path)

print(cv2.__version__)

4.5.1


## **1. Drawing shapes on images**

To draw a line into an image, we'll have to overwrite the image using the `line()` method, in which the first argument is the image itself, the second argument is the coordinate of point1, the third argument is the coordinate of point 2, the fourth argument is the color, and the fifth and last argument is the color.

In [2]:
img = cv2.imread('lena.jpg', 1)
img = cv2.line(img, (0, 0), (255, 255), (0, 0, 255), 5) # Color in BGR format

cv2.imshow('image-with-line', img)

cv2.waitKey(0)
cv2.destroyAllWindows()

We can also make an arrowed line with the `arrowedLine()` method which has the same arguments, taking into consideration that the arrow will be pointing in the direction of the second point.

In [3]:
img = cv2.imread('lena.jpg', 1)
img = cv2.arrowedLine(img, (100, 80), (255, 255), (230, 30, 130), 5) # Color in BGR format

cv2.imshow('image-with-line', img)

cv2.waitKey(0)
cv2.destroyAllWindows()

For drawing rectangles we can use the `rectangle()` method, in which we have the same first five arguments (img, point1(top left), point2(bottom right), color, thickness, ...) and then we can give a sixth argument that would be the line type and the last argument is the shift.

In [4]:
img = cv2.imread('lena.jpg', 1)
img = cv2.arrowedLine(img, (100, 80), (235, 235), (230, 30, 130), 5) # Color in BGR format
img = cv2.rectangle(img, (245, 245), (350, 350), (230, 130, 80), 5)

cv2.imshow('image-with-line', img)

cv2.waitKey(0)
cv2.destroyAllWindows()

And we can also make some differente preferences as filling the recangle (value of -1 in the thickness argument).

In [5]:
img = cv2.imread('lena.jpg', 1)
img = cv2.arrowedLine(img, (100, 80), (235, 235), (230, 30, 130), 5) # Color in BGR format
img = cv2.rectangle(img, (245, 245), (350, 350), (230, 130, 80), -1)

cv2.imshow('image-with-line', img)

cv2.waitKey(0)
cv2.destroyAllWindows()

For drawing circles we use the `circles()` method which takes the image as the first argument, the second argument is the center of the circule, the third argument is the radius, and the 4th adn 5th arguments are the color and thickness respectively.

The complete set of methods it's on the [OpenCV Documentation > Drawing functions](https://docs.opencv.org/3.4/d6/d6e/group__imgproc__draw.html)

In [6]:
img = cv2.imread('lena.jpg', 1)
img = cv2.arrowedLine(img, (100, 80), (235, 235), (230, 30, 130), 5) # Color in BGR format
img = cv2.circle(img, (270, 270), 30, (230, 130, 80), 4)

cv2.imshow('image-with-line', img)

cv2.waitKey(0)
cv2.destroyAllWindows()

## **2. Adding text into images**

Once again we'll overwrite our image with the `putText()` method. Once again the first argument is the image, then the text we want to display, as the 3rd argument the starting point of the text we'll display, the 4th argument is the font face (given by a variable that we'll assign with `.FONT_`). Then, the 5th argument is the font size, the 6th argument is the font color, the 7th is the thickness, and the last one is the linea type (fiven by `LINE_`).

In [7]:
img = cv2.imread('lena.jpg', 1)
img = cv2.circle(img, (270, 270), 30, (230, 130, 80), 4)

fontface = cv2.FONT_HERSHEY_SIMPLEX
img = cv2.putText(img, 'Circle in the eye', (150, 220), fontface, 1, (230, 130, 80), 2, cv2.LINE_AA)

cv2.imshow('image-with-line', img)

cv2.waitKey(0)
cv2.destroyAllWindows()

## **3. Creating images with numpy**

Also, instead of importing an existing image we can create an image by defining a three-dimensional matrix (in the case of color scale image) or just a matrix (for gray-scale images).

For example with `np.zeros()` method in which we can give the height, width and the scale (3 is colored), and then the data type.

In [8]:
import numpy as np

img = np.zeros([512, 512, 3], np.uint8)
img = cv2.circle(img, (270, 270), 30, (230, 130, 80), 4)

fontface = cv2.FONT_HERSHEY_SIMPLEX
img = cv2.putText(img, 'Just a circle', (172, 220), fontface, 1, (230, 130, 80), 2, cv2.LINE_AA)

cv2.imshow('image-with-line', img)

cv2.waitKey(0)
cv2.destroyAllWindows()

## **4. Setting camera parameters**

We can use the `cap.set()` method to set some values. Generally we can set all the properties that we can read with the `cap.get()` method.

In [9]:
cap = cv2.VideoCapture(0)
print('Frame original width: ', cap.get(cv2.CAP_PROP_FRAME_WIDTH))
print('Frame original height: ', cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        

cap.set(cv2.CAP_PROP_FRAME_WIDTH, 200)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 200)

aux = 0

while (cap.isOpened()):
    ret, frame = cap.read()
    cv2.imshow('Video Capture', frame)
    if aux == 0:
        print('Frame set width: ', cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        print('Frame set height: ', cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        aux += 1
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

Frame original width:  640.0
Frame original height:  480.0
Frame set width:  320.0
Frame set height:  240.0


However, the camera acts accordingly to it's resolution. The minimun resoltion is 320 pixels width. What would happen if we set the frame resolution to 300x300?

In [10]:
cap = cv2.VideoCapture(0)
print('Frame original width: ', cap.get(cv2.CAP_PROP_FRAME_WIDTH))
print('Frame original height: ', cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        

cap.set(cv2.CAP_PROP_FRAME_WIDTH, 351)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 270)

aux = 0

while (cap.isOpened()):
    ret, frame = cap.read()
    cv2.imshow('Video Capture', frame)
    if aux == 0:
        print('Frame set width: ', cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        print('Frame set heigth: ', cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        aux += 1
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

Frame original width:  640.0
Frame original height:  480.0
Frame set width:  352.0
Frame set heigth:  288.0


I didn't find in the documentation why the width is taken down from 300 (in the set method) to 288, but my personal conclusion is that, in order to mantain the ratio between width and height, both values must be integer and also, not all values are posible (i.e. only widths multiple of 32 are possible).

Also, the maximum frame size is limited by the maximum resolution of the default camera. For example:

In [11]:
cap = cv2.VideoCapture(0)
print('Frame original width: ', cap.get(cv2.CAP_PROP_FRAME_WIDTH))
print('Frame original height: ', cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

cap.set(cv2.CAP_PROP_FRAME_WIDTH, 2000)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 2000)

aux = 0

while (cap.isOpened()):
    ret, frame = cap.read()
    cv2.imshow('Video Capture', frame)
    if aux == 0:
        print('Frame set width: ', cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        print('Frame set height: ', cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        aux += 1
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

Frame original width:  640.0
Frame original height:  480.0
Frame set width:  1280.0
Frame set height:  720.0


## **5. Showing text in a video**

### **5.1 Showing frame width and height**
First of all we'll be showing the frame width and height within the video we are capturing.

In [12]:
cap = cv2.VideoCapture(0)
print('Frame original width: ', cap.get(cv2.CAP_PROP_FRAME_WIDTH))
print('Frame original height: ', cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 2000)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 2000)

aux = 0

while (cap.isOpened()):
    ret, frame = cap.read()
    
    font = cv2.FONT_HERSHEY_SIMPLEX
    text_w = 'Frame width: ' + str(round(cap.get(cv2.CAP_PROP_FRAME_WIDTH)))
    text_h = 'Frame height: ' + str(round(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
    
    frame = cv2.putText(frame, text_w, (10, 50), font, 1, (255, 255, 255), 2, cv2.LINE_AA)
    frame = cv2.putText(frame, text_h, (10, 100), font, 1, (255, 255, 255), 2, cv2.LINE_AA)
    
    cv2.imshow('Video Capture', frame)
    
    if aux == 0:
        print(text_w)
        print(text_h)
        aux += 1
        
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

Frame original width:  640.0
Frame original height:  480.0
Frame width: 1280
Frame height: 720


### **5.2 Showing date and time in a video**

For this, we'll be using the `datetime` library.

In [13]:
import datetime

cap = cv2.VideoCapture(0)

cap.set(cv2.CAP_PROP_FRAME_WIDTH, 2000)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 2000)

while (cap.isOpened()):
    ret, frame = cap.read()
    
    font = cv2.FONT_HERSHEY_SIMPLEX
    datet = str(datetime.datetime.now().strftime('%H:%M:%S %d/%b/%Y'))
    
    frame = cv2.putText(frame, datet, (10, 40), font, 1, (255, 255, 255), 2, cv2.LINE_AA)
    
    cv2.imshow('Video Capture', frame)
        
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

## **6. Handling mouse events**

He can handle different mouse events such as left click, double left click, right clicks, etc. We can see all the available events within OpenCV by doing:

In [14]:
for i in dir(cv2):
      if 'EVENT' in i:
          print(i)


EVENT_FLAG_ALTKEY
EVENT_FLAG_CTRLKEY
EVENT_FLAG_LBUTTON
EVENT_FLAG_MBUTTON
EVENT_FLAG_RBUTTON
EVENT_FLAG_SHIFTKEY
EVENT_LBUTTONDBLCLK
EVENT_LBUTTONDOWN
EVENT_LBUTTONUP
EVENT_MBUTTONDBLCLK
EVENT_MBUTTONDOWN
EVENT_MBUTTONUP
EVENT_MOUSEHWHEEL
EVENT_MOUSEMOVE
EVENT_MOUSEWHEEL
EVENT_RBUTTONDBLCLK
EVENT_RBUTTONDOWN
EVENT_RBUTTONUP


First of all, we'll create a mouse callback function that will take place when the mouse event occurs, with the help of the `setMouseCallback()` method.

The first argument of the method is the window name (that should be the same everywhere), and as a second argument the callback function.

In [15]:
def click_event(event, x, y, flags, param):
    if event == cv2.EVENT_LBUTTONDOWN:
        print('X: ', x,'   y: ',  y)
        font = cv2.FONT_HERSHEY_SIMPLEX
        coord_str = str(x) + ', ' + str(y)
        cv2.putText(img, coord_str, (x,y), font, 0.5, (0, 0, 0), 1)
        cv2.imshow('image', img)
    if event == cv2.EVENT_RBUTTONDOWN:
        blue = img[y, x, 0]
        green = img[y, x, 1]
        red = img[y, x, 2]
        BGR = '(' + str(blue) + ', ' + str(green) + ', ' + str(red) + ')'
        font = cv2.FONT_HERSHEY_SIMPLEX
        cv2.putText(img, BGR, (x,y), font, 0.5, (30, 100, 210), 1)
        cv2.imshow('image', img)

img = np.ones((512, 512, 3), np.uint8)*255
cv2.imshow('image', img)

cv2.setMouseCallback('image', click_event)

cv2.waitKey(0)
cv2.destroyAllWindows()

X:  211    y:  146
X:  251    y:  280


Maybe, it's more visual if we do this in a real photo. Also, let's change the right click text so it's the opposite color of the pixel selected.

In [16]:
def click_event(event, x, y, flags, param):
    
    if event == cv2.EVENT_LBUTTONDOWN:
        
        print('X: ', x,'   y: ',  y)
        
        font = cv2.FONT_HERSHEY_SIMPLEX
        coord_str = str(x) + ', ' + str(y)
        
        cv2.putText(img, coord_str, (x,y), font, 0.5, (0, 0, 0), 1)
        cv2.imshow('image', img)
        
    if event == cv2.EVENT_RBUTTONDOWN:
        
        blue = img[y, x, 0]
        green = img[y, x, 1]
        red = img[y, x, 2]
        
        BGR = '(' + str(blue) + ', ' + str(green) + ', ' + str(red) + ')'
        font = cv2.FONT_HERSHEY_SIMPLEX
        
        cv2.putText(img, BGR, (x,y), font, 0.5, (255 - int(blue), 255 - int(green), 255 - int(red)), 1)
        cv2.imshow('image', img)

img = cv2.imread('lena.jpg')
cv2.imshow('image', img)

cv2.setMouseCallback('image', click_event)

cv2.waitKey(0)
cv2.destroyAllWindows()

X:  31    y:  65
X:  304    y:  188
X:  274    y:  303


Now we'll draw small circles (points) and when the second cirle is drawn, a line should appear connecting both circles.

In [17]:
def click_event(event, x, y, flags, param):
    
    if event == cv2.EVENT_LBUTTONDOWN:
        
        cv2.circle(img, (x, y), 5, (0, 0, 255), -1)
        points.append((x, y))
        
        if len(points) >= 2:
            cv2.line(img, points[-1], points[-2], (255, 0, 0), 2)
        
        cv2.imshow('image', img)

img = np.zeros((512, 512, 3), np.uint8)
cv2.imshow('image', img)
points = []

cv2.setMouseCallback('image', click_event)

cv2.waitKey(0)
cv2.destroyAllWindows()